### ISSUE TITLE
Unexpected unbound variable with loop, nested if, and attribute usage

### ISSUE THREAD

**Issue by tbhartman** [microsoft/pyright#7217]
**Describe the bug**

Pyright fails to determine that a variable is bound with a combination of syntax:
  - a loop has nested logic which conditionally continues the loop
  - an attribute of the variable is accessed

**Code or Screenshots**

<details><summary>
This is a minimum working example, derived from a poor-man's parser
</summary>

The original parser has a regex in place of the `line == 'a'` test, and `match` is the result of another regex's `match` (so is `re.Match[str]|None`).  If `match is not None`, the `owners` list is appended with one of the matches groups.  However, I've managed to remove the `re` dependency to the point where the example may not look realistic.  For the curious, here is the original, with only the regex's redacted:

```py
import dataclasses
import pathlib
import re
import textwrap

@dataclasses.dataclass
class Result:
    owners:list[str]
    description:str

def parse(p: pathlib.Path):
    if not p.exists() or not p.is_file():
        return
    with open(p) as f:
        BEGIN = re.compile(r'REDACTED')
        OWNER = re.compile(r'REDACTED')
        OWNER_SPLIT = re.compile(r"[\W,]+")
        END = re.compile(r'REDACTED')
        owners: list[str] | None = None
        description = ""
        match = None
        for line in f.readlines():
            if owners is None:
                if BEGIN.match(line):
                    owners = []
                continue
            if END.match(line):
                break
            match = OWNER.match(line)
            if match is not None:
                owners.extend(filter(None, OWNER_SPLIT.split(match.group(1))))
                continue
            description += line.lstrip("#")
        if owners is None:
            return
        description = textwrap.dedent(description)
        return Result(owners, description)
```

In this case, Pyright reports:

    ./parser.py
      ./parser.py:30:16 - error: "match" is unbound (reportUnboundVariable)
      ./parser.py:31:62 - error: "match" is unbound (reportUnboundVariable)

My fix is to use the first workaround described below:

    owners = [] if BEGIN.match(line) else None

</details>

```python
lines:list[str] = list(map(str, range(20)))
owners: list[str] | None = None
for line in lines:
    if owners is None:
        if line == 'a':
            owners = []
        continue
    match = line
    print(match)
    print(owners or match.encode())
    print(owners or match)
```

*Note:* even adding `match:str = ''` before the `for` loop does not help.

***Workarounds***

Any of the following changes allow Pyright to pass:
- change the `if line == 'a'` logic to `owners = [] if line == 'a' else None`
- moving the last four lines into an `else` block and removing the `continue`
- removing only the line with `match.encode`

**VS Code extension or command-line**

With the above code as `test.py`, running Pyright from the command line:

    $ pyright test.py
    test.py
      test.py:9:11 - error: "match" is unbound (reportUnboundVariable)
      test.py:10:21 - error: "match" is unbound (reportUnboundVariable)
      test.py:10:27 - error: Cannot access member "encode" for type "Unbound" (reportAttributeAccessIssue)
      test.py:11:21 - error: "match" is unbound (reportUnboundVariable)
    4 errors, 0 warnings, 0 informations

Pyright version:

    > pyright --version
    pyright 1.1.350

Mypy has no issues with this file:

    > mypy test.py
    Success: no issues found in 1 source file

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem, and it indeed looks like a bug. I'll investigate further.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.352, which I just published.
